File: 2023/day13/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseInput(text) {
    return text.trim().split('\n\n').map(block => block.split('\n'));
}

function findVerticalReflection(pattern) {
    if (!pattern.length) return 0;
    const width = pattern[0].length;

    for (let col = 1; col < width; col++) {
        let isReflection = true;
        for (const row of pattern) {
            const left = row.slice(0, col).split('').reverse().join('');
            const right = row.slice(col);
            const minLen = Math.min(left.length, right.length);
            if (left.slice(0, minLen) !== right.slice(0, minLen)) {
                isReflection = false;
                break;
            }
        }
        if (isReflection) return col;
    }
    return 0;
}

function findHorizontalReflection(pattern) {
    if (!pattern.length) return 0;
    const height = pattern.length;

    for (let row = 1; row < height; row++) {
        let isReflection = true;
        const top = pattern.slice(0, row).reverse();
        const bottom = pattern.slice(row);
        const minLen = Math.min(top.length, bottom.length);
        for (let i = 0; i < minLen; i++) {
            if (top[i] !== bottom[i]) {
                isReflection = false;
                break;
            }
        }
        if (isReflection) return row;
    }
    return 0;
}

function summarizePattern(pattern) {
    const v = findVerticalReflection(pattern);
    if (v > 0) return v;
    return findHorizontalReflection(pattern) * 100;
}

function part1(patterns) {
    return patterns.reduce((sum, p) => sum + summarizePattern(p), 0);
}

function countDifferences(s1, s2) {
    let diff = 0;
    const minLen = Math.min(s1.length, s2.length);
    for (let i = 0; i < minLen; i++) {
        if (s1[i] !== s2[i]) diff++;
    }
    return diff;
}

function findVerticalReflectionWithSmudge(pattern) {
    if (!pattern.length) return 0;
    const width = pattern[0].length;

    for (let col = 1; col < width; col++) {
        let totalDiff = 0;
        for (const row of pattern) {
            const left = row.slice(0, col).split('').reverse().join('');
            const right = row.slice(col);
            const minLen = Math.min(left.length, right.length);
            totalDiff += countDifferences(left.slice(0, minLen), right.slice(0, minLen));
            if (totalDiff > 1) break;
        }
        if (totalDiff === 1) return col;
    }
    return 0;
}

function findHorizontalReflectionWithSmudge(pattern) {
    if (!pattern.length) return 0;
    const height = pattern.length;

    for (let row = 1; row < height; row++) {
        let totalDiff = 0;
        const top = pattern.slice(0, row).reverse();
        const bottom = pattern.slice(row);
        const minLen = Math.min(top.length, bottom.length);
        for (let i = 0; i < minLen; i++) {
            totalDiff += countDifferences(top[i], bottom[i]);
            if (totalDiff > 1) break;
        }
        if (totalDiff === 1) return row;
    }
    return 0;
}

function summarizePatternWithSmudge(pattern) {
    const v = findVerticalReflectionWithSmudge(pattern);
    if (v > 0) return v;
    return findHorizontalReflectionWithSmudge(pattern) * 100;
}

function part2(patterns) {
    return patterns.reduce((sum, p) => sum + summarizePatternWithSmudge(p), 0);
}

const inputFile = join(__dirname, '..', 'input.txt');
const text = readFileSync(inputFile, 'utf-8');
const patterns = parseInput(text);

console.log(`Part 1: ${part1(patterns)}`);
console.log(`Part 2: ${part2(patterns)}`);

File: 2023/day13/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class Solution {

    static class Pattern {
        List<String> lines;

        Pattern(List<String> lines) {
            this.lines = lines;
        }
    }

    public static List<Pattern> parseInput(String text) {
        List<Pattern> patterns = new ArrayList<>();
        String[] blocks = text.trim().split("\n\n");

        for (String block : blocks) {
            String[] lines = block.split("\n");
            List<String> lineList = new ArrayList<>();
            for (String line : lines) {
                lineList.add(line);
            }
            patterns.add(new Pattern(lineList));
        }

        return patterns;
    }

    public static int findVerticalReflection(Pattern pattern) {
        if (pattern.lines.isEmpty()) {
            return 0;
        }

        int width = pattern.lines.get(0).length();

        for (int col = 1; col < width; col++) {
            boolean isReflection = true;

            for (String row : pattern.lines) {
                // Get left side (reversed) and right side
                String left = new StringBuilder(row.substring(0, col)).reverse().toString();
                String right = row.substring(col);

                // Compare overlapping parts
                int minLen = Math.min(left.length(), right.length());
                if (!left.substring(0, minLen).equals(right.substring(0, minLen))) {
                    isReflection = false;
                    break;
                }
            }

            if (isReflection) {
                return col;
            }
        }

        return 0;
    }

    public static int findHorizontalReflection(Pattern pattern) {
        if (pattern.lines.isEmpty()) {
            return 0;
        }

        int height = pattern.lines.size();

        for (int row = 1; row < height; row++) {
            boolean isReflection = true;

            // Get top (reversed) and bottom
            List<String> top = new ArrayList<>();
            for (int i = row - 1; i >= 0; i--) {
                top.add(pattern.lines.get(i));
            }

            List<String> bottom = new ArrayList<>();
            for (int i = row; i < height; i++) {
                bottom.add(pattern.lines.get(i));
            }

            int minLen = Math.min(top.size(), bottom.size());
            for (int i = 0; i < minLen; i++) {
                if (!top.get(i).equals(bottom.get(i))) {
                    isReflection = false;
                    break;
                }
            }

            if (isReflection) {
                return row;
            }
        }

        return 0;
    }

    public static int summarizePattern(Pattern pattern) {
        int v = findVerticalReflection(pattern);
        if (v > 0) {
            return v;
        }
        int h = findHorizontalReflection(pattern);
        return h * 100;
    }

    public static int part1(List<Pattern> patterns) {
        int sum = 0;
        for (Pattern p : patterns) {
            sum += summarizePattern(p);
        }
        return sum;
    }

    public static int countDifferences(String s1, String s2) {
        int count = 0;
        int minLen = Math.min(s1.length(), s2.length());
        for (int i = 0; i < minLen; i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                count++;
            }
        }
        return count;
    }

    public static int findVerticalReflectionWithSmudge(Pattern pattern) {
        if (pattern.lines.isEmpty()) {
            return 0;
        }

        int width = pattern.lines.get(0).length();

        for (int col = 1; col < width; col++) {
            int totalDiff = 0;

            for (String row : pattern.lines) {
                String left = new StringBuilder(row.substring(0, col)).reverse().toString();
                String right = row.substring(col);
                int minLen = Math.min(left.length(), right.length());

                totalDiff += countDifferences(left.substring(0, minLen), right.substring(0, minLen));

                if (totalDiff > 1) {
                    break;
                }
            }

            if (totalDiff == 1) {
                return col;
            }
        }

        return 0;
    }

    public static int findHorizontalReflectionWithSmudge(Pattern pattern) {
        if (pattern.lines.isEmpty()) {
            return 0;
        }

        int height = pattern.lines.size();

        for (int row = 1; row < height; row++) {
            int totalDiff = 0;

            List<String> top = new ArrayList<>();
            for (int i = row - 1; i >= 0; i--) {
                top.add(pattern.lines.get(i));
            }

            List<String> bottom = new ArrayList<>();
            for (int i = row; i < height; i++) {
                bottom.add(pattern.lines.get(i));
            }

            int minLen = Math.min(top.size(), bottom.size());
            for (int i = 0; i < minLen; i++) {
                totalDiff += countDifferences(top.get(i), bottom.get(i));
                if (totalDiff > 1) {
                    break;
                }
            }

            if (totalDiff == 1) {
                return row;
            }
        }

        return 0;
    }

    public static int summarizePatternWithSmudge(Pattern pattern) {
        int v = findVerticalReflectionWithSmudge(pattern);
        if (v > 0) {
            return v;
        }
        int h = findHorizontalReflectionWithSmudge(pattern);
        return h * 100;
    }

    public static int part2(List<Pattern> patterns) {
        int sum = 0;
        for (Pattern p : patterns) {
            sum += summarizePatternWithSmudge(p);
        }
        return sum;
    }

    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../input.txt");
        String text = Files.readString(inputPath);
        List<Pattern> patterns = parseInput(text);

        System.out.println("Part 1: " + part1(patterns));
        System.out.println("Part 2: " + part2(patterns));
    }
}
